worker/autowithdraw.js

```diff
@@ -185,5 +169,5 @@ async function autowithdrawCLN ({ amount, maxFee }, { me, models, lnd }) {
     expiry: 360
   })
 
-  return await createWithdrawal(null, { invoice: inv.bolt11, maxFee }, { me, models, lnd, autoWithdraw: true })
+  return await createWithdrawal(null, { invoice: inv.bolt11, maxFee }, { me, models, lnd, walletId: wallet.id })
```

Comment:
@ekzyis is this omission intentional?

Reply:
Yes since the `createWithdrawal` call in pages/api/lnwith.js is not about an autowithdrawal

Reply:
I see. So once we ship this we'll remove the old procedure and this code will use the new one.

Reply:
Just a heads up, because I'm not entirely sure, but without a default for `wallet_id` I believe `api/pages/lnwith.js` may need to change anyway. But this might also cause conflicts according to [the overloading section here](https://www.postgresql.org/docs/current/sql-createfunction.html).

```diff
+ CREATE OR REPLACE FUNCTION create_withdrawl(lnd_id TEXT, invoice TEXT, msats_amount BIGINT, msats_max_fee BIGINT, username TEXT, auto_withdraw BOOLEAN, wallet_id INTEGER DEFAULT NULL)
- CREATE OR REPLACE FUNCTION create_withdrawl(lnd_id TEXT, invoice TEXT, msats_amount BIGINT, msats_max_fee BIGINT, username TEXT, auto_withdraw BOOLEAN, wallet_id INTEGER)
```

Reply:
Old code calls old signature `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)` (still exists since I didn't drop this signature). New code calls new signature `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN, INTEGER)`

With a default for the last parameter in the new signature they would conflict, yes.

I don't understand why it would need to change without a default though?

I'll test this by running the old code with the new function signature existing, brb

Reply:
Ok, I did the following now:

1. checked out 1faf309c on `master`
2. added the new function to the database manually:

```
stackernews=#  \df "create_withdrawl"
                                                                                  List of functions
 Schema |       Name       | Result data type |                                                      Argument data types                                                      | Type
--------+------------------+------------------+-------------------------------------------------------------------------------------------------------------------------------+------
 public | create_withdrawl | "Withdrawl"      | lnd_id text, invoice text, msats_amount bigint, msats_max_fee bigint, username text, auto_withdraw boolean                    | func
 public | create_withdrawl | "Withdrawl"      | lnd_id text, invoice text, msats_amount bigint, msats_max_fee bigint, username text, auto_withdraw boolean, wallet_id integer | func
(2 rows)
```

4. funded wallet
5. withdraw from wallet which called `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)`

no errors

so I think this is actually fine as it is :eyes:

Reply:
I was mostly concerned about what would happen when the old function is removed in a future commit. I suspect it breaks `api/pages/lnwith.js` if there isn't a default on `wallet_id`.

Reply:
Mhh, I see. So it's about when we remove `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)` and old code is still calling it during the migration. This will fail since that function no longer exists. It wouldn't fail when there would be a default.

Therefore, I see following options:

a) Remove `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)` and replace `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN, INTEGER)` with `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN, INTEGER DEFAULT NULL)` in the same future migration
b) Simply delete `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)` since this will only break withdrawals while we're migrating which shouldn't take too long. This isn't backwards incompatible in a way that it will take down the site afaict. It's still an interesting problem to think about though.
c) Think more about versioned schemas like `create_withdrawal_vX` etc.
d) What are the reasons we're using postgres functions instead of inline JS code? Mostly historical since support for [interactive transactions](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions) in Prisma was only made generally available in 4.7.0? I think there is a case to be made for interactive transactions instead of postgres functions.

Reply:
b sounds fine. Another option

e) have `pages/api/lnwith` use the new version by supplying a null `wallet_id`

Reply:
I think I am super confused _now_ or I _was_ super confused. I don't understand [my own comment](https://github.com/stackernews/stacker.news/pull/1073#discussion_r1566694916) anymore:

> So it's about when we remove `create_withdrawal(TEXT, TEXT, BIGINT, BIGINT, TEXT, BOOLEAN)` and old code is still calling it during the migration

When we deploy the changes in this PR and the currently deployed code (https://github.com/stackernews/stacker.news/commit/1d154e) stops running, there will no longer be code that calls `create_withdrawal(,,,,, BOOLEAN)`. There will only be code from this PR that calls `create_withdrawal(,,,,, BOOLEAN, INTEGER)` including the code in pages/api/lnwith. So we can safely remove `create_withdrawal(,,,,, BOOLEAN)` then.

But we actually will want to migrate from `create_withdrawal(..., BOOLEAN, INTEGER)` to just `create_withdrawal(..., INTEGER)`. Is this the problem you are talking about? Even though I don't see how a `INTEGER DEFAULT NULL` fits in here.

Reply:
I just realized that page/api/lnwith doesn't have an independent call to `create_with..` lol. Let's stop me from confusing you more.

Is this good to merge?

Reply:
This is as good as it gets